library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -108,8 +109,27 @@ compileTerm (Types.Ann _usage ty t) =
     Types.LamM as body -> compileLam ty as body
     Types.AppM fn arguments -> compileApp ty fn arguments
     Types.Var sym -> compileVar sym
+    Types.ArrayIndex index ->
+      compileIndex ty index
+    Types.Closure cap arg body ->
+      compileClosure ty cap arg body
     Types.Prim _t -> mkPrim t ty
 
+-- | @compileTermForApplication@ like @compileTerm@ however it will
+-- promote lambdas to closures to make the calling convention for HOF proper.
+compileTermForApplication ::
+  Types.Define m => Types.Annotated Types.TermClosure -> m LLVM.Operand
+compileTermForApplication ann@(Types.Ann _usage ty t) =
+  case t of
+    Types.LamM as body -> do
+      lam <- compileLam ty as body
+      -- WARNING ∷ Leaking memory
```

Comment:
It might be worth noting in the "Leaking memory" comments that what will close the leak is to implement GC and replace the `malloc` here with the GC's new allocation routine.